example_client/benchmark/metrics.py

```diff
@@ -0,0 +1,352 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import time
+import math
+import re
+
+class XMetrics(dict):
+    def __init__(self, init_data=None, submetrics=1):
+        if init_data is None: init_data = {}
+        dict.__init__(self, init_data)
+        self["submetrics"] = submetrics
+        self.__mean_counter_cache = {}
+        self.__mean_sum_cache = {}
+
+    def calc_sum(self, results, key):
+        if key not in results: return
+        if key not in self: self[key] = 0
+        self[key] += results[key]
+
+    def calc_max(self, results, key):
+        if key not in results: return
+        if key not in self:
+            self[key] = results[key]
+        elif self[key] < results[key]:
+            self[key] = results[key]
+
+    def calc_min(self, results, key):
+        if key not in results: return
+        if key not in self:
+            self[key] = results[key]
+        elif self[key] > results[key]:
+            self[key] = results[key]
+
+    def calc_mean(self, results, key):
+        if key not in results: return
+        if key not in self:
+            self.__mean_counter_cache[key] = 0
+            self.__mean_sum_cache[key] = 0.0
+        self.__mean_counter_cache[key] += 1
+        self.__mean_sum_cache[key] += results[key]
+        self[key] = self.__mean_sum_cache[key] / self.__mean_counter_cache[key]
+
+    def __iadd_series(self, prefix, other):
+        self.calc_sum(other, f"{prefix}total_batches")
+        self.calc_sum(other, f"{prefix}total_frames")
+        self.calc_sum(other, f"{prefix}pass_batches")
+        self.calc_sum(other, f"{prefix}fail_batches")
+        self.calc_sum(other, f"{prefix}pass_frames")
+        self.calc_sum(other, f"{prefix}fail_frames")
+
+        self.calc_sum(other, f"{prefix}netto_batch_rate")
+        self.calc_sum(other, f"{prefix}netto_frame_rate")
+        self.calc_mean(other, f"{prefix}mean_latency")
+        self.calc_max(other, f"{prefix}pass_max_latency")
+        self.calc_max(other, f"{prefix}fail_max_latency")
+
+        self.calc_min(other, f"{prefix}start_timestamp")
+        self.calc_max(other, f"{prefix}stop_timestamp")
+
+        self[f"{prefix}total_duration"] = self[f"{prefix}stop_timestamp"] - self[f"{prefix}start_timestamp"]
+        self[f"{prefix}brutto_batch_rate"] = self[f"{prefix}pass_batches"] / self[f"{prefix}total_duration"]
+        self[f"{prefix}brutto_frame_rate"] = self[f"{prefix}pass_frames"] / self[f"{prefix}total_duration"]
+
+        self.calc_mean(other, f"{prefix}batch_passrate")
+        self.calc_mean(other, f"{prefix}frame_passrate")
+        self.calc_mean(other, f"{prefix}pass_mean_latency")
+        self.calc_mean(other, f"{prefix}fail_mean_latency")
+
+        N = self["submetrics"]
```

Comment:
Is it possible this value to be 0? Because later in this file you divide by N

Reply:
assert is added:

assert N > 1, "need at least 1 client/concurrency"